### PR TITLE
use correct feature flag for impl-block-level trait bounds on const fn

### PR DESCRIPTION
**Pull request by RalfJung**
I am not sure what that special hack was needed for, but it doesn't seem needed any more...

This removes the last use of the `const_fn` feature flag -- Cc https://github.com/rust-lang/rust/issues/84510
r? @oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
Maybe we don't have tests for the situation the comment was referring to? Check if we have a test where a generic impl const block can use its trait bounds in a function, if not, please add one

**Comment by RalfJung** [comment]
I don't understand what that situation would look like, could you sketch an example?

**Comment by oli-obk** [comment]
On phone, but i'll try:

```rust
impl<T: Foo> const Bar for () {
    fn bar() {
        T::foo()
    } 
}
```

Declare the Foo and Bar traits appropriately.

**Comment by RalfJung** [comment]
I did that, but I am not sure if that is the test the comment was referring to... the comment sounded more like a situation where a `const Trait` bound somewhere was incorrectly not recognized as `const` and hence, I assume, leads to an error -- IOW, the comment referred to some code incorrectly failing to build (not some code incorrectly building).

`impl<T: const Plus> const Plus for Num<T>` does not even parse, though.
This code does parse and, to my surprise, fails to build:
```
mod num {
    pub trait Plus {
        fn plus(self, rhs: Self) -> Self;
    }

    struct Num<T>(T);
    impl<T: ?const Plus + Copy> const Plus for Num<T> {
        fn plus(self, rhs: Self) -> Self {
            Num(self.0.plus(rhs.0))
        }
    }
}
```
However, the error has nothing to do with feature gating:
```
error[E0015]: calls in constant functions are limited to constant functions, tuple structs and tuple variants
  --> /home/r/src/rust/rustc.2/src/test/ui/rfc-2632-const-trait-impl/call-generic-method-nonconst-opt-out.rs:30:17
   |
LL |             Num(self.0.plus(rhs.0))
   |                 ^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```
So, I don't think what this PR does makes things any worse here.

**Comment by RalfJung** [comment]
I think what is left here is to fix `lock_api` -- I'll submit a PR there tomorrow (once https://github.com/rust-lang/rust/pull/84578 is in nightly).

**Comment by oli-obk** [comment]
> This code does parse and, to my surprise, fails to build:

Why are you surprised it fails to build? It should fail, you are using `?const` to opt out of constness after all

**Comment by oli-obk** [comment]
The problem is that it also fails to compile without `?const`

**Comment by RalfJung** [comment]
> Why are you surprised it fails to build? It should fail, you are using ?const to opt out of constness after all

Ah, I forgot that things are implicitly `const`.^^

**Comment by RalfJung** [comment]
> The problem is that it also fails to compile without ?const

Okay, I guess I should remove my new test then? But it says nothing about a feature gate there, so I don't think my PR changes the situation here... it might mean some people need to add `feature(const_fn_trait_bounds)` but that does not seem unreasonable?

File: compiler/rustc_mir/src/transform/check_consts/validation.rs [hunk 1]
```diff
@@ -440,16 +440,7 @@ impl Validator<'mir, 'tcx> {
...
-                                {
-                                    // HACK: We shouldn't need the conditional above, but trait
-                                    // bounds on containing impl blocks are wrongly being marked as
-                                    // "not-const".
-                                    self.check_op_spanned(ops::ty::TraitBound(kind), span);
-                                }
+                                self.check_op_spanned(ops::ty::TraitBound(kind), span);
                             }
                             // other kinds of bounds are either tautologies
                             // or cause errors in other passes
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/check_consts/validation.rs:446]: Either the playground is overloaded or

```rust
#![feature(const_trait_impl)]
#![allow(incomplete_features)]

trait Foo<T> {
    fn foo();
}

impl const Foo<u32> for i32 {
    fn foo() {}
}

const fn foo<T>() where i32: Foo<T> {
    <i32 as Foo<T>>::foo()
}
```

hits this path and puts the compiler in an infinite loop.

**Comment by oli-obk** [comment]
> > The problem is that it also fails to compile without ?const
> 
> Okay, I guess I should remove my new test then? But it says nothing about a feature gate there, so I don't think my PR changes the situation here... it might mean some people need to add `feature(const_fn_trait_bounds)` but that does not seem unreasonable?

we already have a test for this behaviour and possibly even an issue, idk where though

File: compiler/rustc_mir/src/transform/check_consts/validation.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_consts/validation.rs:446]: The playground doesn't have my patch though, so how would this indicate that removing the hack is wrong...?

Comment (RalfJung) [reply compiler/rustc_mir/src/transform/check_consts/validation.rs:446]: Also FWIW, [this builds fine for me](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=34c15ed741dbe268841c7954997252ca).

**Comment by RalfJung** [comment]
> we already have a test for this behaviour and possibly even an issue, idk where though

Okay so I removed the new test again. Should I add anything else instead?

**Comment by oli-obk** [comment]
> > we already have a test for this behaviour and possibly even an issue, idk where though
> 
> Okay so I removed the new test again. Should I add anything else instead?

idk, I don't have the time to really dig into this, so let's just merge as is. It preserves the status quo, so all is good

**Comment by RalfJung** [comment]
Cc https://github.com/Amanieu/parking_lot/pull/281

**Pull request merged**